### PR TITLE
Container Hierarch V2

### PR DESCRIPTION
**Pull request by murlock**
Instead of using placeholder, directories are now stored in `redis`.

Done:
- [x] Basic operations GET/PUT/DELETE
- [x] Key is removed when last object is removed
- [x] Segments use same mecanism
- [x] Listing with recursive or not and prefix
- [x] Marker with prefix
- [x] Add Travis test
- [x] Test with distcp
- [x] Fix unit test with `4.2.x` branch
- [x] Optimize redis usage (scan instead of keys)
- [ ] Accept versioning

### CODE REVIEW

**Comment by jkasarherou** [comment]
I want to try this.

File: oioswift/common/middleware/container_sharding.py [hunk 1]
```diff
@@ -0,0 +1,534 @@
...
+
+        LOG.debug("%s: remove key %s (key %s)", self.SWIFT_SOURCE, path, key)
+        res = self.conn.delete(key)
+        if not res:
+            LOG.warn("%s: failed to remove key %s", self.SWIFT_SOURCE, key)
+
+    def _build_object_listing(self, start_response, env,
+                              account, container, prefix,
+                              limit=None,
+                              recursive=False, marker=None):
```
Comment (fvennetier) [inline oioswift/common/middleware/container_sharding.py]: This method is very long, could we split it?

**Comment by fvennetier** [comment]
Also, it looks like a lot of code has been copypasted from `oioswift/common/middleware/container_hierarchy.py`. Either an intermediate class should be created to host this common code, or the previous middleware should be deprecated.

File: oioswift/common/middleware/container_sharding.py (hunk 1 above)
Comment (murlock) [reply oioswift/common/middleware/container_sharding.py]: Will be done in dedicated PR

File: oioswift/common/middleware/container_sharding.py [hunk 2]
```diff
@@ -0,0 +1,534 @@
...
+            else:
+                limit = int(limit[0])
+            container2 = container
+            obj2 = obj
+        else:
+            LOG.debug("%s: -> is NOT listing request", self.SWIFT_SOURCE)
+            obj_parts = obj.split(self.DELIMITER)
+            if len(obj_parts) > 1:
+                path = self.DELIMITER.join(obj_parts[:-1]) + self.DELIMITER
+                is_dir = obj.endswith('/')
```
[the remark this answers is not in the document]
Comment (murlock) [reply oioswift/common/middleware/container_sharding.py]: DONE

File: oioswift/common/middleware/container_sharding.py [hunk 3]
```diff
@@ -0,0 +1,534 @@
...
+            prefix = ['']
+        marker = qs.get('marker')
+        limit = qs.get('limit')
+        LOG.debug("%s: Got %s request for container=%s, "
+                  "obj=%s, prefix=%s marker=%s",
+                  self.SWIFT_SOURCE, req.method, container, obj, prefix,
+                  marker)
+        must_recurse = False
+
+        if 'Oio-Copy-From' in req.headers and req.method == 'PUT':
```
[the remark this answers is not in the document]
Comment (murlock) [reply oioswift/common/middleware/container_sharding.py]: Will be done in dedicated PR

File: oioswift/common/middleware/container_sharding.py [hunk 4]
```diff
@@ -0,0 +1,534 @@
...
+
+        if redis_host or sentinel_hosts:
+            self.conn = RedisDb(
+                redis_host=redis_host,
+                sentinel_hosts=sentinel_hosts,
+                sentinel_name=sentinel_name)
+        else:
+            self.conn = FakeRedis()
+
+    def check_pipeline(self, conf):
```
[the remark this answers is not in the document]
Comment (murlock) [reply oioswift/common/middleware/container_sharding.py]: Will be done in dedicated PR

**Pull request merged**